reco_utils/recommender/sar/sar_singlenode.py

```diff
@@ -301,20 +304,17 @@ def score(self, test, remove_seen=False, normalize=False):
 
         # calculate raw scores with a matrix multiplication
         logger.info("Calculating recommendation scores")
-        # TODO: only compute scores for users in test
-        test_scores = self.user_affinity.dot(self.item_similarity)
-
-        # remove items in the train set so recommended items are always novel
-        if remove_seen:
-            logger.info("Removing seen items")
-            test_scores[self.seen_items[:, 0], self.seen_items[:, 1]] = -np.inf
-
-        test_scores = test_scores[user_ids, :]
+        test_scores = self.user_affinity[user_ids, :].dot(self.item_similarity)
 
         # ensure we're working with a dense ndarray
         if isinstance(test_scores, sparse.spmatrix):
             test_scores = test_scores.toarray()
 
+        # remove items in the train set so recommended items are always novel
+        if remove_seen:
+            logger.info("Removing seen items")
+            test_scores = test_scores * self.seen_items[user_ids, :]
```

Comment:
here we can add the reduced user_affinity back in after multiplying with -inf, zeros will stay zero and anything we consider the user has seen will have a non-zero entry which will turn to -inf.
```
test_scores += self.user_affinity[user_ids, :] * -np.inf
```